src/rust/engine/src/tasks.rs

```diff
@@ -45,43 +45,133 @@ impl rule_graph::Rule for Rule {
       &Rule::Intrinsic(_) => false,
     }
   }
+
+  fn color(&self) -> Option<rule_graph::Palette> {
+    match self {
+      Rule::Task(_) => None,
+      Rule::Intrinsic(_) => Some(rule_graph::Palette::Gray),
+    }
+  }
+
+  fn fmt_for_graph(&self) -> String {
+    match self {
+      Rule::Task(ref task) => {
+        let FormattedTaskRuleElements {
+          rule_type,
+          task_name,
+          clause_portion,
+          product,
+          get_portion,
+        } = Self::extract_task_elements(
+          task,
+          Some(GraphVisualizationParameters {
+            select_clause_threshold: 2,
+            get_clause_threshold: 1,
+          }),
+        );
+
+        format!(
+          "@{}({}) -> {}{}\n{}",
+          rule_type, clause_portion, product, get_portion, task_name,
+        )
+      }
+      Rule::Intrinsic(ref intrinsic) => format!(
+        "@rule(<intrinsic>({}) -> {})",
+        intrinsic.input, intrinsic.product,
+      ),
+    }
+  }
+}
+
+///
+/// A helper struct to contain stringified versions of various components of the rule.
+///
+struct FormattedTaskRuleElements {
+  rule_type: String,
+  task_name: String,
+  clause_portion: String,
+  product: String,
+  get_portion: String,
+}
+
+#[derive(Clone, Copy)]
+struct GraphVisualizationParameters {
+  select_clause_threshold: usize,
```

Comment:
I would add a comment on both members of this struct because I don't think the name gives the full context of what they represent.